drivers/counter/counter_handlers.c

```diff
@@ -11,38 +11,138 @@
  * instance and return an integral value
  */
 #define COUNTER_HANDLER(name) \
-	static inline int z_vrfy_counter_##name(struct devince *dev) \
+	static inline int z_vrfy_counter_##name(struct device *dev) \
 	{ \
 		Z_OOPS(Z_SYSCALL_DRIVER_COUNTER(dev, name)); \
 		return z_impl_counter_ ## name((struct device *)dev); \
 	}
 
 COUNTER_HANDLER(get_pending_int)
-COUNTER_HANDLER(read)
 COUNTER_HANDLER(stop)
 COUNTER_HANDLER(start)
-COUNTER_HANDLER(get_top_value)
-COUNTER_HANDLER(get_max_relative_alarm)
 
 #include <syscalls/counter_get_pending_int_mrsh.c>
-#include <syscalls/counter_read_mrsh.c>
 #include <syscalls/counter_stop_mrsh.c>
 #include <syscalls/counter_start_mrsh.c>
+
+static inline bool z_vrfy_counter_is_counting_up(const struct device *dev)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_COUNTER));
+	return z_impl_counter_is_counting_up((const struct device *)dev);
+}
+#include <syscalls/counter_is_counting_up_mrsh.c>
+
+static inline u8_t z_vrfy_counter_get_num_of_channels(const struct device *dev)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_COUNTER));
+	return z_impl_counter_get_num_of_channels((const struct device *)dev);
+}
+#include <syscalls/counter_get_num_of_channels_mrsh.c>
+
+static inline u32_t z_vrfy_counter_get_frequency(const struct device *dev)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_COUNTER));
+	return z_impl_counter_get_frequency((const struct device *)dev);
+}
+#include <syscalls/counter_get_frequency_mrsh.c>
+
+static inline u32_t z_vrfy_counter_us_to_ticks(const struct device *dev,
+					       u64_t us)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_COUNTER));
+	return z_impl_counter_us_to_ticks((const struct device *)dev,
+					  (u64_t)us);
+}
+#include <syscalls/counter_us_to_ticks_mrsh.c>
+
+static inline u64_t z_vrfy_counter_ticks_to_us(const struct device *dev,
+					       u32_t ticks)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_COUNTER));
+	return z_impl_counter_ticks_to_us((const struct device *)dev,
+					  (u32_t)ticks);
+}
+#include <syscalls/counter_ticks_to_us_mrsh.c>
+
+static inline u32_t z_vrfy_counter_read(struct device *dev)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_COUNTER(dev, read));
+	return z_impl_counter_read((struct device *)dev);
+}
+#include <syscalls/counter_read_mrsh.c>
+
+static inline int z_vrfy_counter_set_channel_alarm(struct device *dev,
+			u8_t chan_id, const struct counter_alarm_cfg *alarm_cfg)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_COUNTER(dev, set_alarm));
+	Z_OOPS(Z_SYSCALL_MEMORY_READ((const struct counter_alarm_cfg *)
```

Comment:
You have to make a copy of the alarm_cfg structure before passing it to the implementation, to prevent TOCTOU attacks.

Something like:

```
struct counter_alarm_cfg cfg_copy;
Z_OOPS(z_user_from_copy(&cfg_copy, alarm_cfg, sizeof(cfg_copy));
```

And then do all the remaining checks using cfg_copy, passing its address to the z_impl function. The z_user_from_copy() logic verifies that the source memory is readable so you won't need a separate Z_SYSCALL_MEMORY_READ() call.

Reply:
@andrewboie Thank you. Fixed.